ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,960 @@
+    /// See [`Uds::connect_network()`] to connect to a network or [`Uds::create_network()`] to create one.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// let channel = uds.get_channel()?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsGetChannel")]
+    pub fn get_channel(&self) -> Result<u8, Error> {
+        if self.service_status() == ServiceStatus::Disconnected {
+            return Err(Error::NotConnected);
+        }
+
+        let mut channel = MaybeUninit::uninit();
+
+        ResultCode(unsafe { ctru_sys::udsGetChannel(channel.as_mut_ptr()) })?;
+
+        let channel = unsafe { channel.assume_init() };
+
+        Ok(channel)
+    }
+
+    /// Wait for a ConnectionStatus event to occur.
+    ///
+    /// If `next` is `true`, discard the current event (if any) and wait for the next one.
+    ///
+    /// If `wait` is `true`, block until an event is signalled, else return `false` if no event.
+    ///
+    /// Always returns `true`, unless `wait` is `false` and no event has been signalled.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the service is currently neither connected to nor hosting a network.
+    /// See [`Uds::connect_network()`] to connect to a network or [`Uds::create_network()`] to create one.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// if uds.wait_status_event(false, false)? {
+    ///     println!("Event signalled");
+    /// }
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsWaitConnectionStatusEvent")]
+    pub fn wait_status_event(&self, next: bool, wait: bool) -> Result<bool, Error> {
+        if self.service_status() == ServiceStatus::Disconnected {
+            return Err(Error::NotConnected);
+        }
+
+        Ok(unsafe { ctru_sys::udsWaitConnectionStatusEvent(next, wait) })
+    }
+
+    /// Returns the current [`ctru_sys::udsConnectionStatus`] struct.
+    ///
+    /// TODO: should this return an error if not connected?
```

Comment:
It seems like calling it when not connected actually returns success. On New 3DS, I get
```rs
ConnectionStatus {
    status: 3,
    unk_x4: 0,
    cur_node_id: 2,
    unk_xa: 0,
    unk_xc: [0, 0, 0, 0, 0, 0, 0, 0],
    total_nodes: 0,
    max_nodes: 0,
    node_bitmask: 0,
}
```
and on Old 3DS I get
```rs
ConnectionStatus {
    status: 3,
    unk_x4: 0,
    cur_node_id: 0,
    unk_xa: 0,
    unk_xc: [0, 0, 0, 0, 0, 0, 0, 0],
    total_nodes: 0,
    max_nodes: 0,
    node_bitmask: 0,
}
```
consistently (they're identical besides `cur_node_id`).

Reply:
Interesting. At this point I wonder whether the `status` member itself specifies whether or not the client is connected to a network. We should probably try to understand its meaning.

Reply:
`libctru` doesn't describe it at all, so it would very much be a reverse-engineering effort. From my 20 seconds of experimentation, it looks like a value of 6 represents connected. Presumably it's a bitfield.

Reply:
Found some details, seems like it's more of a weird enum with sparse value choices? https://www.3dbrew.org/wiki/NWMUDS:GetConnectionStatus#Status_values

Reply:
The `TODO` doc comment should be removed now that it's fixed.